server/src/DBTablesLastInfo.cc

```diff
@@ -0,0 +1,483 @@
+	arg.add(IDX_LAST_INFO_DATA_TYPE, lastInfoDef.dataType);
+	arg.add(IDX_LAST_INFO_VALUE, lastInfoDef.value);
+	arg.add(IDX_LAST_INFO_SERVER_ID, lastInfoDef.serverId);
+
+	getDBAgent().runTransaction(arg);
+	return HTERR_OK;
+}
+
+HatoholError DBTablesLastInfo::getLastInfoList(LastInfoDefList &lastInfoDefList,
+					       const LastInfoQueryOption &option)
+{
+
+	DBAgent::SelectExArg arg(tableProfileLastInfo);
+	arg.add(IDX_LAST_INFO_ID);
+	arg.add(IDX_LAST_INFO_DATA_TYPE);
+	arg.add(IDX_LAST_INFO_VALUE);
+	arg.add(IDX_LAST_INFO_SERVER_ID);
+
+	// condition
+	arg.condition = option.getCondition();
+
+	getDBAgent().runTransaction(arg);
+
+	const ItemGroupList &grpList = arg.dataTable->getItemGroupList();
+	ItemGroupListConstIterator itemGrpItr = grpList.begin();
+	for (; itemGrpItr != grpList.end(); ++itemGrpItr) {
+		ItemGroupStream itemGroupStream(*itemGrpItr);
+		LastInfoDef lastInfoDef;
+
+		itemGroupStream >> lastInfoDef.id;
+		itemGroupStream >> lastInfoDef.dataType;
+		itemGroupStream >> lastInfoDef.value;
+		itemGroupStream >> lastInfoDef.serverId;
+
+		lastInfoDefList.push_back(lastInfoDef);
+	}
+
+	return HTERR_OK;
+}
+
+static string makeIdListCondition(const LastInfoIdList &idList)
+{
+	string condition;
+	const ColumnDef &colId = COLUMN_DEF_LAST_INFO[IDX_LAST_INFO_ID];
+	condition = StringUtils::sprintf("%s in (", colId.columnName);
+	LastInfoIdListConstIterator it = idList.begin();
+	while (true) {
+		const int id = *it;
+		condition += StringUtils::sprintf("%d", id);
+		++it;
+		if (it == idList.end())
+			break;
+		condition += ",";
+	}
+	condition += ")";
+	return condition;
+}
+
+static string makeServerIdCondition(string serverId)
+{
+	return StringUtils::sprintf(
+		 "%s=%" FMT_LAST_INFO_SERVER_ID,
+		 COLUMN_DEF_LAST_INFO[IDX_LAST_INFO_SERVER_ID].columnName,
+		 serverId.c_str());
+}
+static string makeConditionForDelete(const LastInfoIdList &idList,
+				     const OperationPrivilege &privilege)
+{
+	string condition = makeIdListCondition(idList);
+
+	return condition;
+}
+
+HatoholError DBTablesLastInfo::deleteLastInfoList(const LastInfoIdList &idList,
+                                                  const OperationPrivilege &privilege)
+{
+	HatoholError err = checkPrivilegeForDelete(privilege, idList);
+	if (err != HTERR_OK)
+		return err;
+
+	if (idList.empty()) {
+		MLPL_WARN("idList is empty.\n");
+		return HTERR_INVALID_PARAMETER;
+	}
+
+	struct TrxProc : public DBAgent::TransactionProc {
+		DBAgent::DeleteArg arg;
+		uint64_t numAffectedRows;
+
+		TrxProc (void)
+		: arg(tableProfileLastInfo),
+		  numAffectedRows(0)
+		{
+		}
+
+		void operator ()(DBAgent &dbAgent) override
+		{
+			dbAgent.deleteRows(arg);
+			numAffectedRows = dbAgent.getNumberOfAffectedRows();
+		}
+	} trx;
+	trx.arg.condition = makeConditionForDelete(idList, privilege);
+	getDBAgent().runTransaction(trx);
+
+	// Check the result
+	if (trx.numAffectedRows != idList.size()) {
```

Comment:
Is this check really needed ?
How about just showing a warning ?
In the current architecture, the number of the affected should be the same as the number of requested.
However, we could add the mechanism that deletes records concurrently.